add CellCast tensor function

arnej27959 opened the pull request:
* with GenericCellCast instruction implementation
* and reference operation

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@havardpe please review

arnej27959 committed:
add CellCast tensor function

* with GenericCellCast instruction implementation
* and reference operation

havardpe commented on eval/src/vespa/eval/eval/test/reference_operations.cpp:
```diff
@@ -72,6 +72,18 @@ vespalib::string rename_dimension(const vespalib::string &name, const std::vecto
+        return result;
+    }
+    for (const auto & [ addr, value ]: a.cells()) {
+        result.add(addr, value);
```
we should restrict values here to match the abilities of 'to'

arnej27959 committed:
changes after review

* add ValueType::cell_cast which computes the canonical result type
  when applying a cast operator.  For now, only allow double as scalar.
* slight change to make_instruction; add the cell type as an extra
  parameter and require consistency (future-proofing).
* In reference implementation, copy values via a temporary of
  the appropriate type, ensuring that casts may lose precision
  like they should.

havardpe commented on eval/src/vespa/eval/eval/value_type.cpp:
```diff
@@ -388,6 +388,12 @@ ValueType::either(const ValueType &one, const ValueType &other) {
+ValueType
+ValueType::cell_cast(const ValueType &from, CellType to_cell_type) {
+    // XXX we want to change this and use make_type later:
+    return tensor_type(from.dimensions(), to_cell_type);
```
I think we should do make_type here already and rather make type-resolving decide what is legal or not

havardpe commented on eval/src/tests/instruction/generic_cell_cast/generic_cell_cast_test.cpp:
```diff
@@ -37,8 +37,8 @@ TensorSpec perform_generic_cell_cast(const TensorSpec &a, CellType to, const Val
-    auto res_type = ValueType::make_type(to, lhs->type().dimensions());
-    auto my_op = GenericCellCast::make_instruction(res_type, lhs->type(), stash);
+    auto res_type = ValueType::cell_cast(lhs->type(), to);
+    auto my_op = GenericCellCast::make_instruction(res_type, lhs->type(), res_type.cell_type(), stash);
```
prefer passing 'to' directly to make_instruction

havardpe commented on eval/src/vespa/eval/instruction/generic_cell_cast.cpp:
```diff
@@ -43,9 +43,11 @@ struct SelectGenericCellCastOp {
 InterpretedFunction::Instruction
 GenericCellCast::make_instruction(const ValueType &result_type,
                                   const ValueType &input_type,
+                                  CellType to_cell_type,
```
we should be able to trust ValueType::cell_cast and not have to pass the result_type

arnej27959 committed:
review followup

* assume type resolving takes case of any special-cases
  regarding cell_cast operations on scalar types.

arnej27959 committed:
use cell type from result as final cast target
